### PR TITLE
Add ContextVar-backed CuprumContext with scoped allowlists and hooks

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds a complete CuprumContext system backed by ContextVar, with scoped allowlists, before/after hooks, and deterministic ordering. Includes API surface, tests (unit, behaviour, and features), and documentation updates to reflect usage and design decisions.

### Core functionality
- Implemented cuprum/context.py introducing:
  - CuprumContext: immutable, ContextVar-backed context carrying an allowlist, before_hooks, and after_hooks.
  - Narrowing semantics:
    - When parent allowlist is empty, provided allowlist becomes the new base.
    - When parent allowlist is non-empty, provided allowlist is intersected (narrowing only).
  - Hook handling with deterministic ordering:
    - Before hooks execute in registration order (FIFO).
    - After hooks execute in reverse order (LIFO).
  - Context management:
    - scoped() context manager for entering a narrowed context.
    - _ScopedContext facilitating safe enter/exit with restoration.
  - Registration helpers:
    - AllowRegistration: dynamic extension and detachment of allowlists.
    - HookRegistration: registration and detachment for before/after hooks.
  - Utility API:
    - current_context(), get_context(), with_allowlist(), with_before_hook(), with_after_hook(), etc.
  - Exceptions:
    - ForbiddenProgramError raised when checking/entering a non-allowed program.
  - Type aliases for BeforeHook and AfterHook.

### API exposure
- Updated cuprum/__init__.py to export new context-related symbols:
  - AfterHook, BeforeHook, CuprumContext, ForbiddenProgramError, HookRegistration
  - after, before, allow, scoped, current_context, get_context

### Tests
- Added unit tests: cuprum/unittests/test_context.py
  - Verifies basic context behavior, narrowing semantics, hook ordering, and error handling.
- Added behavioural tests: tests/behaviour/test_context_hooks.py
  - Covers scoped narrowing, hook registration/detachment, and thread/async task isolation.
- Added feature file: tests/features/context_hooks.feature
  - Documents scenarios for context hooks and scoped behavior.
- Expanded design/test coverage to ensure deterministic hook ordering and safe context restoration across nested scopes and concurrent execution.

### Documentation
- Updated docs/cuprum-design.md with a dedicated CuprumContext Decisions section explaining:
  - Allowlist narrowing semantics
  - Hook ordering (FIFO for before, LIFO for after)
  - ContextVar usage and immutability guarantees
  - Registration and detachment semantics
- Updated docs/roadmap.md to reflect completion of the context/hook scope work
- Expanded docs/users-guide.md with:
  - Scoped contexts usage examples
  - Dynamic allowlist extension via allow()
  - Before/after hook usage examples and ordering
  - Thread and async task isolation notes

## Testing plan
- Run the full test suite locally:
  - pytest -q
  - pytest tests/behaviour -q
  - pytest cuprum/unittests/test_context.py -q
- Verify:
  - Scoped contexts correctly narrow and restore across nested blocks.
  - Before hooks run in registration order; after hooks run in reverse order.
  - Dynamic allowlist extension via AllowRegistration and detach() behaves correctly.
  - HookRegistration can be used as a context manager and detaches on exit.
  - Context isolation across threads and async tasks is preserved.
  - ForbiddenProgramError is raised for disallowed programs and passed for allowed ones.

## Usage highlights
- Scoped contexts:
-  from cuprum import ECHO, scoped
+  with scoped(allowlist=frozenset([ECHO])) as ctx:
+      assert ctx.is_allowed(ECHO)
+      # embed nested scopes to narrow further

- Dynamic allowlisting:
+  from cuprum import LS, allow, current_context, scoped
+  with scoped(allowlist=frozenset([ECHO])):
+      reg = allow(LS)
+      assert current_context().is_allowed(LS)
+      reg.detach()
+      assert not current_context().is_allowed(LS)
+
- Hooks:
+  from cuprum import before, after
+  with scoped(allowlist=frozenset([ECHO])):
+      with before(log_before), after(log_after):
+          # commands executed will trigger hooks
+
+### Breaking changes
+- None expected. The new CuprumContext API is additive and designed to be backward-compatible with existing usage patterns.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---

ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/2dec09a9-6212-4091-ad25-8870a2e8ff68

### CODE REVIEW

File: docs/users-guide.md [hunk 1]
```diff
@@ -155,3 +155,186 @@ def greet() -> None:
...
+
+def log_after(cmd, result):
+    print(f"Finished {cmd.program} with exit code {result.exit_code}")
+
+
+with scoped(allowlist=frozenset([ECHO])):
+    with before(log_before), after(log_after):
+        cmd = sh.make(ECHO)("hello")
+        # Hooks will be invoked when cmd.run() is called
+```
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:269]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md:269]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: cuprum/unittests/test_context.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    assert ctx.after_hooks == ()
+
+
+def test_context_with_hooks() -> None:
+    """Context retains provided hooks."""
+    before: BeforeHook = mock.Mock()
+    after: AfterHook = mock.Mock()
+    ctx = CuprumContext(before_hooks=(before,), after_hooks=(after,))
+    assert ctx.before_hooks == (before,)
+    assert ctx.after_hooks == (after,)
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_context.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: cuprum/context.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+
+    """
+
+    allowlist: frozenset[Program] = dc.field(default_factory=frozenset)
+    before_hooks: tuple[BeforeHook, ...] = ()
+    after_hooks: tuple[AfterHook, ...] = ()
+
+    def is_allowed(self, program: Program) -> bool:
+        """Return True when the program is in the allowlist."""
+        return program in self.allowlist
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/context.py:66]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: cuprum/context.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        _set_context(new_ctx)
+
+    def detach(self) -> None:
+        """Remove the allowed programs from the current context."""
+        if self._detached:
+            return
+        self._detached = True
+        ctx = current_context()
+        new_ctx = dc.replace(ctx, allowlist=ctx.allowlist - self._programs)
+        _set_context(new_ctx)
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/context.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: cuprum/context.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+        """Remove the hook from the current context."""
+        if self._detached:
+            return
+        self._detached = True
+        ctx = current_context()
+        if self._hook_type == "before":
+            new_ctx = ctx.without_before_hook(typ.cast("BeforeHook", self._hook))
+        else:
+            new_ctx = ctx.without_after_hook(typ.cast("AfterHook", self._hook))
+        _set_context(new_ctx)
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/context.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: cuprum/unittests/test_context.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+def test_scoped_restores_on_exception() -> None:
+    """scoped() restores context even when exception is raised."""
+    original = current_context()
+    with (
+        pytest.raises(ValueError),  # noqa: PT011
+        scoped(allowlist=frozenset([ECHO])),
+    ):
+        raise ValueError("test")
+    assert current_context() is original
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_context.py:170]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: cuprum/unittests/test_safe_cmd_run.py [hunk 7]
```diff
@@ -326,3 +326,161 @@ def _poll_process_death(pid: int, *, timeout: float = 1.0) -> None:
...
+
+    _, execute = execution_strategy
+    command = sh.make(ECHO)("hello")
+    # Create a context with an allowlist that does NOT include ECHO
+    other_program = Program("cat")
+    with (
+        scoped(allowlist=frozenset([other_program])),
+        pytest.raises(ForbiddenProgramError),
+    ):
+        execute(command, {})
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_safe_cmd_run.py:351]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

**Comment by leynos** [comment]
@coderabbitai review
